x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/AbstractLookupService.java

```diff
@@ -0,0 +1,586 @@
+                        return loader;
+                    }
+                )
+            );
+        }
+        return new ValuesSourceReaderOperator(
+            driverContext.blockFactory(),
+            fields,
+            List.of(new ValuesSourceReaderOperator.ShardContext(searchContext.searcher().getIndexReader(), searchContext::newSourceLoader)),
+            0
+        );
+    }
+
+    private Page createNullResponse(int positionCount, List<NamedExpression> extractFields) {
+        final Block[] blocks = new Block[extractFields.size()];
+        try {
+            for (int i = 0; i < extractFields.size(); i++) {
+                blocks[i] = blockFactory.newConstantNullBlock(positionCount);
+            }
+            return new Page(blocks);
+        } finally {
+            if (blocks[blocks.length - 1] == null) {
+                Releasables.close(blocks);
+            }
+        }
+    }
+
+    private class TransportHandler implements TransportRequestHandler<T> {
+        @Override
+        public void messageReceived(T request, TransportChannel channel, Task task) {
+            request.incRef();
+            ActionListener<LookupResponse> listener = ActionListener.runBefore(new ChannelActionListener<>(channel), request::decRef);
+            doLookup(
+                request,
+                (CancellableTask) task,
+                listener.delegateFailureAndWrap(
+                    (l, outPage) -> ActionListener.respondAndRelease(l, new LookupResponse(outPage, blockFactory))
+                )
+            );
+        }
+    }
+
+    abstract static class Request {
+        final String sessionId;
+        final String index;
+        final DataType inputDataType;
+        final Page inputPage;
+        final List<NamedExpression> extractFields;
+
+        Request(String sessionId, String index, DataType inputDataType, Page inputPage, List<NamedExpression> extractFields) {
+            this.sessionId = sessionId;
+            this.index = index;
+            this.inputDataType = inputDataType;
+            this.inputPage = inputPage;
+            this.extractFields = extractFields;
+        }
+    }
+
+    abstract static class TransportRequest extends org.elasticsearch.transport.TransportRequest implements IndicesRequest {
+        final String sessionId;
+        final ShardId shardId;
+        final DataType inputDataType;
+        final Page inputPage;
+        final List<NamedExpression> extractFields;
+        // TODO: Remove this workaround once we have Block RefCount
+        final Page toRelease;
+        final RefCounted refs = AbstractRefCounted.of(this::releasePage);
+
+        TransportRequest(String sessionId, ShardId shardId, DataType inputDataType, Page inputPage, List<NamedExpression> extractFields) {
+            this.sessionId = sessionId;
+            this.shardId = shardId;
+            this.inputDataType = inputDataType;
+            this.inputPage = inputPage;
+            this.toRelease = null;
```

Comment:
I think we need to pass toRelease when reading a request from an InputStream. Do we now have sufficient infrastructure to remove this workaround?

Reply:
I'll check on this one!

Reply:
I did it.